### PR TITLE
Feat/xtokens

### PR DESCRIPTION
**Pull request by GopherJ**
close: #214

### CODE REVIEW

**Comment by GopherJ** [comment]
I'm ready for review @yz89 @kaichaosun

**Review by ayushmishra2005 — APPROVED** [review]

**Comment by GopherJ** [comment]
anyway let's merge because I need to work on frontend this afternoon, if there are still something to do I will raise a new PR

File: pallets/liquid-staking/Cargo.toml [hunk 1]
```diff
@@ -11,6 +11,7 @@ targets = ['x86_64-unknown-linux-gnu']
 codec            = { package = 'parity-scale-codec', version = '2.0.0', default-features = false }
 frame-support    = { git = 'https://github.com/paritytech/substrate.git', branch = 'polkadot-v0.9.5', default-features = false, version = '3.0.0' }
 frame-system     = { git = 'https://github.com/paritytech/substrate.git', branch = 'polkadot-v0.9.5', default-features = false, version = '3.0.0' }
+max-encoded-len  = { git = 'https://github.com/paritytech/substrate', branch = 'polkadot-v0.9.5', default-features = false }
```
Comment (kaichaosun) [inline pallets/liquid-staking/Cargo.toml:14]: How is this crate used?

File: pallets/liquid-staking/src/lib.rs [hunk 2]
```diff
@@ -264,11 +270,19 @@ pub mod pallet {
...
+                MultiLocation::X2(
+                    Junction::Parent,
+                    Junction::AccountId32 {
+                        network: NetworkId::Any,
+                        id: agent.clone().into(),
+                    },
+                ),
                 amount,
+                // TODO : measure xcm weight
+                1000_1000,
```
Comment (kaichaosun) [inline pallets/liquid-staking/src/lib.rs:285]: The value looks weird, you probably mean 100_000?

File: runtime/heiko/Cargo.toml [hunk 3]
```diff
@@ -174,4 +177,7 @@ std                = [
   'pallet-authorship/std',
   'pallet-collator-selection/std',
   'cumulus-primitives-timestamp/std',
+  'orml-unknown-tokens/std',
```
Comment (kaichaosun) [inline runtime/heiko/Cargo.toml:180]: How is this unknown token used?

File: runtime/heiko/src/lib.rs [hunk 4]
```diff
@@ -268,6 +271,86 @@ impl orml_currencies::Config for Runtime {
     type WeightInfo = ();
 }
 
+pub struct CurrencyIdConvert;
+impl Convert<CurrencyId, Option<MultiLocation>> for CurrencyIdConvert {
+    fn convert(id: CurrencyId) -> Option<MultiLocation> {
+        match id {
+            CurrencyId::KSM => Some(X1(Parent)),
+            CurrencyId::xKSM => Some(X3(
```
Comment (kaichaosun) [inline runtime/heiko/src/lib.rs:279]: We don't need enable xksm now.

File: runtime/heiko/src/lib.rs [hunk 5]
```diff
@@ -268,6 +271,86 @@ impl orml_currencies::Config for Runtime {
...
+        }
+    }
+}
+
+impl Convert<MultiLocation, Option<CurrencyId>> for CurrencyIdConvert {
+    fn convert(location: MultiLocation) -> Option<CurrencyId> {
+        match location {
+            X1(Parent) => Some(CurrencyId::KSM),
+            X3(Parent, Parachain(id), GeneralKey(key))
+                if ParaId::from(id) == ParachainInfo::parachain_id() && key == b"xKSM".to_vec() =>
```
Comment (kaichaosun) [inline runtime/heiko/src/lib.rs:294]: Same here.

**Pull request merged**